### DISCUSSION TITLE
MPR

### DISCUSSION THREAD

**Discussion by ChaithikaPrabudda** [fo-dicom/fo-dicom#1525] [Q&A]
how can i use this library to create MPR? any documentation?

**Comment by gofal** [comment]
There are only some basic classes, that help you to extract some cuts out of a volume. This part of fo-dicom is something that would need some contributors to enhance it. But this is, what it can so far:

in Namespace `FellowOakDicom.Imaging.Reconstruction` you will find some classes. The main of them is `VolumeData`.

1. you create a VolumeData, by passing in a list of ImageData instances. These are wrappers around DicomDatasets providing all the properties that the VolumeData class needs. This can later also be just some in-memory pixel data.
```
IEnumerable<string> files;
var volume = new VolumeData(files.Select (f => new ImageData(f)));
```
2. This VolumeData then can give you some information like the pixelspacing, the slice distances, the bounding box in patient coordinate system etc. This class also has the method `GetCut` where you can extract a cut and get the raw data as byte array
3. You can also use the helper class `Slice`. If you have a Coordinate in patient Space and two direction vectors then you can extract an Slice-instance. This instance then will provide you some more properties. And the Imaging packackes do have extension methods, to render this slice as System.Drawing.Bitmap or as ImageSharp.Image.
```
// calculate a coronal cut
var topLeft = new Point3D(volume.BoundingMin.X, (volume.BoundingMin.Y + volume.BoundingMax.Y) / 2, volume.BoundingMax.Z);
var rowVector = new Vector3D(1,0,0);
var colVector = new Vector3D(0,0,-1);
var pixelSpacing = 0.5; // 0.5mm per pixel
var numberOfRows = (volume.BoundingMax.Z - volume.BoundingMin.Z) / pixelSpacing;
var numberOfCols = (volume.BoundingMax.X - volume.BoundingMin.X) / pixelSpacing;
var slice = new Slice(volume, topLeft, rowVector, colVector, numberOfRows, numberOfCols, pixelSpacing);
// render with ImageSharp
var image = slice.AsImage();
// render wiht System.Drawing
var image = slice.AsBitmap();
```
4. You also can create a whole stack of images. Therefore you can use the `Stack` class, where you can create sagittal, axial or coronal stacks out of a volume
```
var sliceDistance = 2.0; // 2 mm distance between slices
var sagittalStack = new Stack(volume, StackType.Sagittal, pixelSpacing, sliceDistance);
// now you have an array of slices, that you can render or store as dicoms
foreach (var slice in sagittalStack.Slices) {
   var frame = slice.AsImage();
}
```

There are still some things to do, like having a more flexible way of creating a whole stack.
And currently a Slice is just a thin cut, but if you want do do an MPR, then your slice also requires some "thickness" and you need some calucalation method like "Avg", "MIP", "MinMip". etc.

**Reply by ChaithikaPrabudda** [reply]
thank you for the reply. i have tried this, this is working. but it takes a bit of time in creating a Stack instance. is there a way to optimize it? and does it provide any memory cleaning?

**Reply by ChaithikaPrabudda** [reply]
and one more thing, im not much clear on your following point. can you please explain more. 
"And currently a Slice is just a thin cut, but if you want do do an MPR, then your slice also requires some "thickness" and you need some calucalation method like "Avg", "MIP", "MinMip". etc"

**Reply by gofal** [reply]
Of course there is a lot of space for optimization. As I wrote above this can be taken as a starting point, but there are still some things to do. 

The the calcualtion of the slice, i mean that you have to implement one of the common methods of how to aggretage the voxel data into the planar pixel. In practical you never have that thin cuts, because it would make finding of small spots very hard. You do have some SliceThickness additionl to the SliceDistance, sometimes the SliceThickness larger than the SliceDistance leading to some overlapping of the slices. And when aggregating the voxels, then you can do something like summing them up, having average, choosing the minimum value or choosing the maximum value.
